### DISCUSSION TITLE
Why not "defer" bootstrap javascript in <head>?

### DISCUSSION THREAD

**Discussion by SamuelMiller** [twbs/bootstrap#36462] [Ideas]
With "async" and "defer" options for controlling when javascript is loaded, why does Boostrap documentation still recommend placing the bootstrap <script> tag  before the closing body tag </body> (https://getbootstrap.com/docs/5.2/getting-started/introduction/)? It seems like antiquated advice. The "defer" attribute tells the browser to execute the javascript after the document is loaded and parsed (they wait if needed), right before DOMContentLoaded. According to (https://caniuse.com/script-defer), 97% of all browsers support "defer".  

As stated in this thread (https://stackoverflow.com/questions/56105230/bootstrap-4-scripts-best-location-and-scripts-tags), "Head is for external resources. To have all dependencies out of the way at the start of the document is best practice. The `<head>` was made for exactly that. As the DOM is not loaded at that moment in page loading, this will fail all references to the DOM made there. There used to be a (dark) time, before defer, where it was common to get the advice to place the script-includes after `</body>` but before `</html>`."

This:
```
<head>
...
<script defer src="assets/js/bootstrap.bundle.min.js"></script>
</head>
```

Instead of this:
```
...
<script src="assets/js/bootstrap.bundle.min.js"></script>
</body>
</html>
```

**Comment by mdo** [comment]
Great question! I imagine we can make this change, but I'm curious if there's anything I'm overlooking. Thoughts @XhmikosR?

**Comment by David-Maran** [comment]
It depends on how you want to do it. [Here](https://www.alsacreations.com/astuce/lire/1562-script-attribut-async-defer.html) is an explanation that I find well done (in French but use google trad) with tests and choose yourself.

**Comment by aavmurphy** [comment]
have been using cdn bootstrap js and cdn popper js, both with defer for some years -  works fine

**Reply by gpcaretti** [reply]
it works well for me too, but what about layout shifts due to a deferred load of bootstrap js and css?

**Reply by Cmeem** [reply]
> it works well for me too, but what about layout shifts due to a deferred load of bootstrap js and css?

Deferred CSS? what do you mean?

**Comment by Dev-AbdulHannan** [comment]
Because we want our Html and CSS to load first and javascript at the last 
It will eventually improve the page initial load time

**Comment by Cmeem** [comment]
Any progress on this?

**Reply by ghost** [reply]
The answer is in the conversation, just read.

**Reply by Cmeem** [reply]
> The answer is in the conversation, just read.

There is no answer yet

https://github.com/orgs/twbs/discussions/36462#discussioncomment-2868299

**Reply by David-Maran** [reply]
yes there is a answer, it's not just defer, it's also async, it's up to you, not bootstrap.. And the starter template is just a start.

**Reply by Cmeem** [reply]
> yes there is a answer, it's not just defer, it's also async, it's up to you, not bootstrap.. And the starter template is just a start.

how? async will load the script before dom is created fully (means domcontentloaded). defer seems to be working but the devs did not add any comment for defer.

**Reply by David-Maran** [reply]
Most developers use the sass version not cdn.

**Reply by Cmeem** [reply]
> Most developers use the sass version not cdn.

w t f? Are you bot? what sass CSS file has to do with JavaScript flie?

**Reply by David-Maran** [reply]
Because when you use sass you download the full package. And nice speech by the way, such a little kid.

**Comment by saturdaywalkers** [comment]
Play nicely people.
A lot of people aren't experts and just copy Bootstrap.
`<script defer bootstrap> <script defer my-code>` is the modern way to do it, and saves putting everything inside an on-ready event function
Bootstrap should switch to recommending it

**Comment by Sbrjt** [comment]
https://stackoverflow.com/questions/42543729
